src/client/dfuse/pil4dfs/dfs_dcache.c

```diff
@@ -0,0 +1,663 @@
+{
+	dcache_rec_t *rec;
+
+	rec = dlist2drec(rlink);
+	if (rec->dr_key_len != key_len)
+		return false;
+	D_ASSERT(rec->dr_key[key_len] == '\0');
+	D_ASSERT(((char *)key)[key_len] == '\0');
+
+	return strncmp(rec->dr_key, (const char *)key, key_len) == 0;
+}
+
+static void
+dcache_rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int           rc;
+	dcache_rec_t *rec;
+
+	rec = dlist2drec(rlink);
+	D_DEBUG(DB_TRACE, "delete record " DF_DK " (ref=%u)", DP_DK(rec->dr_key),
+		atomic_load(&rec->dr_ref));
+	rc = dfs_release(rec->dr_obj);
+	if (rc)
+		DS_ERROR(rc, "dfs_release() failed");
+	D_FREE(rec);
+}
+
+static void
+dcache_rec_addref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dcache_rec_t *rec;
+	uint32_t      oldref;
+
+	rec    = dlist2drec(rlink);
+	oldref = atomic_fetch_add(&rec->dr_ref, 1);
+	D_DEBUG(DB_TRACE, "increment ref counter of record " DF_DK " from %u to %u",
+		DP_DK(rec->dr_key), oldref, oldref + 1);
+}
+
+static bool
+dcache_rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dcache_rec_t *rec;
+	uint32_t      oldref;
+
+	rec    = dlist2drec(rlink);
+	oldref = atomic_fetch_sub(&rec->dr_ref, 1);
+	D_DEBUG(DB_TRACE, "decrement ref counter of record " DF_DK " from %u to %u",
+		DP_DK(rec->dr_key), oldref, oldref - 1);
+	D_ASSERT(oldref >= 1);
+	return oldref == 1;
+}
+
+static uint32_t
+dcache_rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	dcache_rec_t *rec;
+
+	rec = dlist2drec(rlink);
+	return d_hash_string_u32(rec->dr_key, rec->dr_key_len);
+}
+
+static d_hash_table_ops_t dcache_hash_ops = {
+    .hop_key_cmp    = dcache_key_cmp,
+    .hop_rec_addref = dcache_rec_addref,
+    .hop_rec_decref = dcache_rec_decref,
+    .hop_rec_free   = dcache_rec_free,
+    .hop_rec_hash   = dcache_rec_hash,
+};
+
+static int
+dcache_destroy_act(dfs_dcache_t *dcache);
+static int
+dcache_find_insert_act(dfs_dcache_t *dcache, char *path, size_t path_len, dcache_rec_t **rec);
+static void
+drec_incref_act(dfs_dcache_t *dcache, dcache_rec_t *rec);
+static void
+drec_decref_act(dfs_dcache_t *dcache, dcache_rec_t *rec);
+static void
+drec_del_at_act(dfs_dcache_t *dcache, dcache_rec_t *rec);
+static int
+drec_del_act(dfs_dcache_t *dcache, char *path, dcache_rec_t *parent);
+
+static inline int
+dcache_add_root(dfs_dcache_t *dcache, dfs_obj_t *obj)
+{
+	dcache_rec_t *rec;
+	int           rc;
+
+	D_ALLOC(rec, sizeof(*rec) + DCACHE_KEY_PREF_SIZE);
+	if (rec == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	rec->dr_obj = obj;
+	atomic_init(&rec->dr_ref, 0);
+	atomic_init(&rec->dr_deleted, false);
+	memcpy(&rec->dr_key_child_prefix[0], &dcache->dd_key_root_prefix[0], DCACHE_KEY_PREF_SIZE);
+	memcpy(&rec->dr_key[0], &dcache->dd_key_root_prefix[0], DCACHE_KEY_PREF_SIZE);
+	rec->dr_key_len = DCACHE_KEY_PREF_SIZE - 1;
+	rc = d_hash_rec_insert(&dcache->dd_dir_hash, rec->dr_key, rec->dr_key_len, &rec->dr_entry,
+			       true);
```

Comment:
Does `exclusive` need to be true here?  If the hash table is empty then presumably it could be false although for the first entry it won't make much difference.

Reply:
We used "true" for "exclusive" originally in pil4dfs code.
Yes. we could use "false" here since the record should be unique.

Reply:
At this point, the hash table shall be empty.  If this record is already present, then we have a consistancy issue as this function should only be called at the creation of the dcache